CRM/MembershipExtras/Hook/Pre/MembershipCreate.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/**
+ * Implements hook to be run before a membership is created/edited.
+ */
+class CRM_MembershipExtras_Hook_Pre_MembershipCreate {
+
+  /**
+   * Parameters that will be used to create the membership.
+   *
+   * @var array
+   */
+  private $params;
+
+  public function __construct(&$params) {
+    $this->params = &$params;
+  }
+
+  /**
+   * Preprocesses parameters used for Membership operations.
+   */
+  public function preProcess() {
+    $this->recalculateTaxAmount();
+    $this->recalculateLineItemsAmounts();
+  }
+
+  /**
+   * Recalculates tax amount tax rate according to selected financial type, as
+   * this can be broken if paying using payment plan or altering total value.
+   */
+  private function recalculateTaxAmount() {
+    $taxRates = CRM_Core_PseudoConstant::getTaxRates();
+    $rate = CRM_Utils_Array::value($this->params['financial_type_id'], $taxRates, 0);
+
+    $this->params['tax_amount'] = ($this->params['total_amount'] * ($rate / 100)) / (1 + ($rate / 100));

```

Comment:
I think you need to use : 

```php
round(AMOUNT, 2)
```
for all amount calculations here so it match how civicrm do it